pkg/tempo/servicegraphprocessor/processor.go

```diff
@@ -0,0 +1,255 @@
+	p.serviceGraphRequestHistogram = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "tempo_service_graph_request_seconds",
+		Help:    "Time for a request between two nodes",
+		Buckets: prometheus.ExponentialBuckets(0.01, 2, 12),
+	}, []string{"client", "server"})
+	p.serviceGraphUnpairedSpansTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "tempo_service_graph_unpaired_spans_total",
+		Help: "Total count of requests between two nodes",
+	}, []string{"client", "server"})
+	p.serviceGraphUntaggedSpansTotal = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "tempo_service_graph_untagged_spans_total",
+		Help: "Total count of spans processed that were not tagged with span.kind",
+	}, []string{"span_kind"})
+
+	cs := []prometheus.Collector{
+		p.serviceGraphRequestTotal,
+		p.serviceGraphRequestFailedTotal,
+		p.serviceGraphRequestHistogram,
+		p.serviceGraphUnpairedSpansTotal,
+		p.serviceGraphUntaggedSpansTotal,
+	}
+
+	for _, c := range cs {
+		if err := p.reg.Register(c); err != nil {
+			return err
+		}
+	}
+
+	// Collect unpaired spans when evicting items from the store during
+	// periodic cleanup
+	p.store.OnEvicted(func(s string, i interface{}) {
+		e := i.(edge)
+		if !e.complete() {
+			p.serviceGraphUnpairedSpansTotal.WithLabelValues(e.clientService, e.serverService).Inc()
+		}
+	})
+
+	return nil
+}
+
+func (p *processor) Shutdown(context.Context) error {
+	p.unregisterMetrics()
+	return nil
+}
+
+func (p *processor) unregisterMetrics() {
+	cs := []prometheus.Collector{
+		p.serviceGraphRequestTotal,
+		p.serviceGraphRequestFailedTotal,
+		p.serviceGraphRequestHistogram,
+		p.serviceGraphUnpairedSpansTotal,
+		p.serviceGraphUntaggedSpansTotal,
+	}
+
+	for _, c := range cs {
+		p.reg.Unregister(c)
+	}
+}
+
+func (p *processor) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{}
+}
+
+func (p *processor) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	level.Debug(p.logger).Log("msg", "consuming traces")
+
+	var errs error
+	for _, trace := range batchpersignal.SplitTraces(td) {
+		if err := p.consume(trace); err != nil {
+			if errors.Is(err, errTooManyItems) {
+				level.Warn(p.logger).Log("msg", "skipped processing of spans", "maxItems", p.maxItems, "err", errTooManyItems)
+				break
+			}
+			errs = multierror.Append(errs, err)
+		}
+	}
+	if errs != nil {
+		level.Error(p.logger).Log("msg", "failed consuming traces", "err", errs)
+	}
+
+	p.collectMetrics()
+
+	return p.nextConsumer.ConsumeTraces(ctx, td)
+}
+
+func (p *processor) collectMetrics() {
+	for k, v := range p.store.Items() {
+		e := v.Object.(edge)
+		if e.complete() {
+			p.serviceGraphRequestTotal.WithLabelValues(e.clientService, e.serverService).Inc()
+			if e.failed {
+				p.serviceGraphRequestFailedTotal.WithLabelValues(e.clientService, e.serverService).Inc()
+			}
+			p.serviceGraphRequestHistogram.WithLabelValues(e.clientService, e.serverService).Observe(e.serverLatency.Seconds())
```

Comment:
Shouldn't we also observe `e.clientLatency` here?

Reply:
Yeah, I wasn't sure how to approach this one. Maybe having two metrics for client and server latencies.